packages/ni-grpc-extensions/src/ni_grpc_extensions/loggers.py

```diff
@@ -0,0 +1,499 @@
+            # Serilog provides for ASP.NET Core services. Don't log exception details because
+            # grpcio's logging_pool already handles this.
+            elapsed_time = time.perf_counter() - self._start_time
+            _logger.info(
+                "gRPC server call %s responded %s in %.4f ms",
+                self._method_name,
+                str(_get_status_code(exception)).replace("StatusCode.", ""),
+                elapsed_time * 1000.0,
+            )
+        _logger.debug("gRPC server call complete: %s", self._method_name)
+        _tracelogging.log_grpc_server_call_stop(self._method_name, self._activity_id)
+
+    def log_streaming_request(self) -> None:
+        _logger.debug("gRPC server call streaming request: %s", self._method_name)
+        _tracelogging.log_grpc_server_call_streaming_request(self._method_name)
+
+    def log_streaming_response(self) -> None:
+        _logger.debug("gRPC server call streaming response: %s", self._method_name)
+        _tracelogging.log_grpc_server_call_streaming_response(self._method_name)
+
+
+def _get_status_code(exception: BaseException | None) -> grpc.StatusCode:
+    if exception is None:
+        return grpc.StatusCode.OK
+    elif isinstance(exception, grpc.RpcError):
+        return exception.code()
+    else:
+        return grpc.StatusCode.UNKNOWN
+
+
+_T = TypeVar("_T")
+
+
+class _LoggingRequestIterator(Generic[_T]):
+    __slots__ = ["_call_logger", "_inner_iterator"]
+
+    def __init__(self, call_logger: _CallLogger, inner_iterator: Iterator[_T]) -> None:
+        self._call_logger = call_logger
+        self._inner_iterator = inner_iterator
+
+    def __iter__(self) -> Iterator[_T]:
+        return self
+
+    def __next__(self) -> _T:
+        request = next(self._inner_iterator)
+        self._call_logger.log_streaming_request()
+        return request
+
+
+class _LoggingResponseIterator(Generic[_T]):
+    __slots__ = ["_call_logger", "_inner_iterator"]
+
+    def __init__(self, call_logger: _CallLogger, inner_iterator: Iterator[_T]) -> None:
+        self._call_logger = call_logger
+        self._inner_iterator = inner_iterator
+
+    def __iter__(self) -> Iterator[_T]:
+        return self
+
+    def __next__(self) -> _T:
+        # For server-streaming and bidirectional RPCs, the call is complete when the response
+        # stream is closed or it throws an exception.
+        try:
+            response = next(self._inner_iterator)
+            self._call_logger.log_streaming_response()
+            return response
+        except StopIteration:
+            self._call_logger.close()
+            raise
+        except Exception as e:
+            self._call_logger.close(e)
+            raise
+
+
+if TYPE_CHECKING:
+    # These types only exist in grpc-stubs.
+    _CallFuture = grpc._CallFuture  # pyright: ignore
+    _CallIterator = grpc._CallIterator  # pyright: ignore
+else:
```

Comment:
```suggestion
if TYPE_CHECKING:
```suggestion
    # These types are marked as @type_check_only and can be used only in type annotations,
    # so we redefine them as empty classes at run time. Pyright still warns due to the behavior
    # described in https://github.com/microsoft/pyright/issues/9934
    _CallFuture = grpc._CallFuture  # pyright: ignore[reportGeneralTypeIssues]
    _CallIterator = grpc._CallIterator  # pyright: ignore[reportGeneralTypeIssues]  
```
else:
```